src/plugins/hidsim/hid_injection.cpp

```diff
@@ -0,0 +1,759 @@
+ * on https://github.com/tklengyel/drakvuf, or by other methods.           *
+ * By sending these changes, it is understood (unless you specify          *
+ * otherwise) that you are offering unlimited, non-exclusive right to      *
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ * This file was created by Jan Gruber.                                    *
+ * It is distributed as part of DRAKVUF under the same license             *
+ ***************************************************************************/
+
+#include<fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <linux/input.h>
+#include <sys/time.h>
+#include <math.h>
+
+#include <libdrakvuf/libdrakvuf.h> /* eprint_current_time */
+#include "../private.h" /* PRINT_DEBUG */
+
+/* Hidsim specific includes */
+#include "keymap_evdev_to_qapi.h" /* Mapping evdev->qapi name */
+#include "qmp_connection.h"
+#include "hid_injection.h"
+
+/* Drakvuf HID template */
+#define DRAK_MAGIC_NUMBER 0xc4d2c1cb
+#define DRAK_MAGIC_STR 0x4b415244
+#define HEADER_LEN 0xC
+
+#define PPM_MAGIC_NUMBER "P6"
+
+/* Length of the QMP command buffer */
+#define CMD_BUF_LEN 0x400
+
+/* QAPI's button values */
+#define L_BTN_STR "left"
+#define M_BTN_STR "middle"
+#define R_BTN_STR "right"
+
+/* Throttle injection down to 50 microsec intervals */
+#define TIME_BIN 50
+
+struct dimensions
+{
+    int width;      /* display width */
+    int height;     /* display height */
+    float dx;       /* one pixel equivalent on x-axis */
+    float dy;       /* one pixel equivalent on y-axis */
+};
+
+/* Keep track of the current cursor position */
+int new_x = 1 << 14;
+int new_y = 1 << 14;
+
+static int dump_screen(qmp_connection* qc, const char* path)
+{
+    char cmd[0x200];
+    snprintf(cmd, 0x200,
+        QMP_SCREEN_DUMP_FMT_STR,
+        path);
+    char* out = NULL;
```

Comment:
I would really use json-c here instead of substituting path "by hand". This was a source of errors already in the past.

Reply:
Resolved.